test/unit/org/apache/cassandra/concurrent/SimulatedExecutorFactory.java

```diff
@@ -0,0 +1,512 @@
+            LongSupplier large = () -> rs.nextLong(maxSmall, max);
+            this.jitterNanos = Gens.bools().runs(rs.nextInt(1, 11) / 100.0D, rs.nextInt(3, 15)).mapToLong(b -> b ? large.getAsLong() : small.getAsLong()).asLongSupplier(rs);
+        }
+
+        @Override
+        protected <T> RunnableFuture<T> newTaskFor(Runnable runnable, T value)
+        {
+            return new FutureTask<>(Executors.callable(runnable, value));
+        }
+
+        @Override
+        protected <T> RunnableFuture<T> newTaskFor(Callable<T> callable)
+        {
+            return new FutureTask<>(callable);
+        }
+
+        protected org.apache.cassandra.utils.concurrent.RunnableFuture taskFor(Runnable command)
+        {
+            if (command instanceof org.apache.cassandra.utils.concurrent.RunnableFuture)
+                return (org.apache.cassandra.utils.concurrent.RunnableFuture) command;
+            return new FutureTask<>(Executors.callable(command));
+        }
+
+        @Override
+        public void shutdown()
+        {
+            shutdown = true;
+        }
+
+        @Override
+        public List<Runnable> shutdownNow()
+        {
+            // TODO (now, correctness): should this track pending work?
+            return Collections.emptyList();
+        }
+
+        @Override
+        public boolean isShutdown()
+        {
+            return shutdown;
+        }
+
+        @Override
+        public boolean isTerminated()
+        {
+            return shutdown;
+        }
+
+        @Override
+        public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException
+        {
+            return shutdown;
+        }
+
+        @Override
+        public void execute(Runnable command)
+        {
+            checkNotShutdown();
+            queue.add(new Item(nowWithJitter(), SimulatedExecutorFactory.this.seq++, taskFor(command)));
+        }
+
+        protected void checkNotShutdown()
+        {
+            if (isShutdown())
+                throw new RejectedExecutionException("Shutdown");
+        }
+
+        protected long nowWithJitter()
+        {
+            return nanoTime() + jitterNanos.getAsLong();
+        }
+    }
+
+    private class OrderedExecutorService extends UnorderedExecutorService
+    {
+        private final Queue<Item> pending = new LinkedList<>();
+
+        @Override
+        public void execute(Runnable command)
+        {
+            checkNotShutdown();
+            boolean wasEmpty = pending.isEmpty();
+            Item task = new Item(nowWithJitter(), SimulatedExecutorFactory.this.seq++, taskFor(command));
+            pending.add(task);
+            if (wasEmpty)
+                runNextTask();
+        }
+
+        private void runNextTask()
+        {
+            Item next = pending.peek();
+            if (next == null)
+                return;
+
+            next.action.addCallback((s, f) -> afterExecution());
+            queue.add(next);
```

Comment:
So in the ordered case, we're just kind of placing and processing a single item on the queue over and over (from the pending list, simulating submission ordered execution)?

Reply:
yep.  When we have single thread executors then the execution order is FIFO, so having a later on task run first violates expected behavior..